cores/arduino/SoftwareSerial.cpp

```diff
@@ -0,0 +1,420 @@
+    } else {
+      timer.detachInterrupt();
+    }
+    cur_speed = speed;
+  }
+}
+
+// This function sets the current object as the "listening"
+// one and returns true if it replaces another
+bool SoftwareSerial::listen()
+{
+  if (_receivePin >= 0) {
+    // wait for any transmit to complete as we may change speed
+    while (active_out)
+      ;
+    if (active_listener) {
+      active_listener->stopListening();
+    }
+    rx_tick_cnt = 1;
+    rx_bit_cnt = -1;
+    setSpeed(_speed);
+    active_listener = this;
+    if (!_half_duplex) {
+      active_in = this;
+    }
+    return true;
+  }
+  return false;
+}
+
+// Stop listening. Returns true if we were actually listening.
+bool SoftwareSerial::stopListening()
+{
+  if (active_listener == this) {
+    // wait for any output to complete
+    while (active_out)
+      ;
+    if (_half_duplex) {
+      setRXTX(false);
+    }
+    active_listener = nullptr;
+    active_in = nullptr;
+    // turn off ints
+    setSpeed(0);
+    return true;
+  }
+  return false;
+}
+
+inline void SoftwareSerial::setTX()
+{
+  digitalWriteFast(digitalPinToPinName(_transmitPin), _inverse_logic ? LOW : HIGH);
+  pinMode(_transmitPin, OUTPUT);
+}
+
+inline void SoftwareSerial::setRX()
+{
+  if (_receivePin > 0) {
+    pinMode(_receivePin, _inverse_logic ? INPUT_PULLDOWN : INPUT_PULLUP); // pullup for normal logic!
+  }
+}
+
+inline void SoftwareSerial::setRXTX(bool input)
+{
+  if (_half_duplex) {
+    if (input) {
+      if (active_in != this) {
+        setRX();
+        rx_bit_cnt = -1;
+        rx_tick_cnt = 2;
+        active_in = this;
+      }
+    } else {
+      if (active_in == this) {
+        setTX();
+        active_in = nullptr;
+      }
+    }
+  }
+}
+
+inline void SoftwareSerial::send()
+{
+  if (--tx_tick_cnt <= 0) {
+    if (tx_bit_cnt++ < 10) {
+      // send data (including start and stop bits)
+      digitalWriteFast(digitalPinToPinName(_transmitPin), (tx_buffer & 1) ? ~_inverse_logic : _inverse_logic);
+      tx_buffer >>= 1;
+      tx_tick_cnt = OVERSAMPLE;
+    } else {
+      tx_tick_cnt = 1;
+      if (_output_pending) {
+        active_out = nullptr;
+      } else if (tx_bit_cnt > 10 + OVERSAMPLE * 5) {
+        if (_half_duplex && active_listener == this) {
+          // setRXTX(true);
+          pinMode(_receivePin, _inverse_logic ? INPUT_PULLDOWN : INPUT_PULLUP); // pullup for normal logic!
+          rx_bit_cnt = -1;
+          rx_tick_cnt = 2;
+          active_in = this;
+        }
+        active_out = nullptr;
+      }
+    }
+  }
+}
+
+//
+// The receive routine called by the interrupt handler
+//
+inline void SoftwareSerial::recv()
+{
+  if (--rx_tick_cnt <= 0) {
+    //    uint8_t inbit = gpio_get(_receivePin);
+    uint8_t inbit = digitalReadFast(digitalPinToPinName(_receivePin));
+    if (rx_bit_cnt == -1) {
+      // waiting for start bit
+      if (!inbit) {
+        // got start bit
+        rx_bit_cnt = 0;
+        rx_tick_cnt = OVERSAMPLE + 1;
```

Comment:
I guess OVERSAMPLE +1 is to sample next bit in the middle of the bit (more precisely inafter 1/3 of the bit due to OVERSAMPLE =3) ?
A comment is welcome.

Reply:
The rewrite will remove this artifact